onnx/defs/schema.cc

```diff
@@ -822,18 +829,67 @@ const FunctionProto* OpSchema::GetFunction() const {
+  return const_cast<OpSchema*>(this)->GetFunctionWithOpsetInternal(opset_version, domain, fail_on_invalid_op);
 }
 
-FunctionProto* OpSchema::GetFunctionWithOpsetInternal(int opset_version) {
-  std::map<int, std::shared_ptr<FunctionProto>>::iterator it =
-      opset_version_to_function_body_.lower_bound(opset_version);
-  if (it != opset_version_to_function_body_.cend())
-    return it->second.get();
+void OpSchema::ValidateReferencedOpsInFunciton(
+    const FunctionProto* function,
+    int requested_opset_version,
+    int function_since_version,
+    const std::string& domain,
+    bool fail_on_invalid_op) {
+  if (requested_opset_version == function_since_version) {
+    // we validate that OpSchemas optained with requested_opset_version and function_since_version
+    // are the same for ops used in a function definition.
+    return;
+  }
+  std::stringstream err;
+  bool has_invalid_op = false;
+  for (int node_index = 0; node_index < function->node_size(); ++node_index) {
+    const NodeProto& node = function->node(node_index);
+    const OpSchema* op1 = OpSchemaRegistry::Instance()->GetSchema(node.op_type(), requested_opset_version, domain);
+    const OpSchema* op2 = OpSchemaRegistry::Instance()->GetSchema(node.op_type(), function_since_version, domain);
+    if (op1 != op2) {
+      if (!has_invalid_op) {
+        err << "A function op (" << function->name() << ") with opset version (" << requested_opset_version
+            << ") is requested.\n";
+        err << "A function definition of opset since_version (" << function_since_version << ") shall be used.\n";
+        err << "However, there is/are conflict(s) with operators used to define the function: \n";
+      }
+      err << "Operator (" << node.op_type() << ") of version " << requested_opset_version
+          << " is used but the op has been updated at least once since version " << function_since_version << ".\n";
+      has_invalid_op = true;
+    }
+  }
 
-  return nullptr;
+  if (has_invalid_op) {
+    if (fail_on_invalid_op) {
+      fail_schema(err.str());
+    } else {
+      std::cout << err.str();
+    }
+  }
 }
+
+FunctionProto*
+OpSchema::GetFunctionWithOpsetInternal(int opset_version, const std::string& domain, bool fail_on_invalid_op) {
+  std::map<int, std::shared_ptr<FunctionProto>>::iterator it =
+      opset_version_to_function_body_.upper_bound(opset_version);
+  if (opset_version_to_function_body_.empty() || it == opset_version_to_function_body_.begin()) {
+    ONNX_THROW_EX(std::out_of_range(
+        std::string("Cannot find a function body that satisfies the requested opset version: op_type = ") +
+        this->name_ + ", opset_version = " + std::to_string(opset_version) + "."));
+  } else {
+    --it;
+    int function_since_version = it->first;
+    FunctionProto* function = it->second.get();
+    ValidateReferencedOpsInFunciton(function, opset_version, function_since_version, domain, fail_on_invalid_op);
```

Comment:
If validation fails, we should not return function, we should return a nullptr (plus extra status/error message if desired).